### PR TITLE
System tests refactor

### PR DESCRIPTION
**Pull request by juanje**
I'd like to propose a refactor of the system's tests. The main reason is that the tests have dependencies between them and with external conditions. By definition, tests should be able to work isolated and without any order or dependency.

Also, I think that using very common libraries for BATS as [bats-support](https://github.com/ztombol/bats-support) and [bats-assert](https://github.com/ztombol/bats-assert) might make the tests more explicit and easy to maintain (and to contribute).

The helper was a bit complex for this project. It might fix a more complex project like **libpod**, but I think here makes less sense and make it harder to debug.

I implemented the same tests, but using some extra asserts (mostly for exit codes) and a couple of tests more. If you like the tests, I'd like to add some more.

The `README.md` and the Ansible playbooks are updated as well, but it can be improved.

There are also some notes because I found some possible bugs or improvements (at Toolbox) while I was writing the tests.

### CODE REVIEW

**Comment by juanje** [comment]
Some tests have failed due the inconsistency of the registry (or the network), I need to work on that to make it more reliable.
I was looking for ways to mock the repository and use local images, so it doesn't rely on the network.

Pull the images first (in some tests) is an option, but then it doesn't test the **toolbox**'s mechanism to pull the images the first time or when is asked for an image that is not been pulled yet.

**Comment by juanje** [comment]
I added some helpers for pulling the images before using them on tests, but the registry is failing a lot today :disappointed: 
Some errors `503` from the server and errors like this one:
```
Trying to pull registry.fedoraproject.org/f33/fedora-toolbox:33...
Getting image source signatures
Copying blob sha256:e159cba4da71e8f92486a69703e5c387b05070a1d06adfce03f2ae1dc50721b8
Copying blob sha256:782585b12f5b57ac30c39f43740dffcba7e28e9482e8a2c75a5e81547b925ebd
  unexpected EOF
Error: unable to pull registry.fedoraproject.org/f33/fedora-toolbox:33: Error writing blob: error storing blob to file "/var/tmp/storage649780343/2": unexpected EOF
```

**Comment by HarryMichal** [comment]
Hi @juanje! Thank you for working on this. Much appreciated!

> I added some helpers for pulling the images before using them on tests, but the registry is failing a lot today disappointed
> Some errors `503` from the server and errors like this one:
> 
> ```
> Trying to pull registry.fedoraproject.org/f33/fedora-toolbox:33...
> Getting image source signatures
> Copying blob sha256:e159cba4da71e8f92486a69703e5c387b05070a1d06adfce03f2ae1dc50721b8
> Copying blob sha256:782585b12f5b57ac30c39f43740dffcba7e28e9482e8a2c75a5e81547b925ebd
>   unexpected EOF
> Error: unable to pull registry.fedoraproject.org/f33/fedora-toolbox:33: Error writing blob: error storing blob to file "/var/tmp/storage649780343/2": unexpected EOF
> ```

Yes, the `503` errors were my main motivation to make the tests depend on each other because pulling from `registry.fedoraproject.org` is **very** unreliable and causes the CI to flake **a lot**. I think we should contact the registry's maintainers and try to solve this as this is an long-going issue.

> Pull the images first (in some tests) is an option, but then it doesn't test the toolbox's mechanism to pull the images the first
> time or when is asked for an image that is not been pulled yet.

I'm perfectly fine with adding a helper function for pulling the images. Toolbox's image pulling mechanism is simply a wrapper around the `podman pull` command. It is not Toolbox's (or Podman's) fault that `registry.fedoraproject.org` is unreliable.

> Also, I think that using very common libraries for BATS as bats-support and bats-assert might make the tests more explicit
> and easy to maintain (and to contribute).

Definitely, yes! I'm aware of those libraries but I was hesitant to use those because they are not packaged in Fedora (just the original `bats` from which `bats-core` and family were forked.

> I implemented the same tests, but using some extra asserts (mostly for exit codes) and a couple of tests more. If you like
> the tests, I'd like to add some more.

I'd love to see more tests but first we need to figure out solutions for some of the problems we're facing now (unreliable registry, https://github.com/containers/toolbox/issues/410 - this one might be solved thanks to your update).

Also, in the future, we'd like to hook up Toolbox's tests with Podman's. We still have to discuss this matter with Podman developers to get more info about this, though.

Again, thank you for working on this!

**Comment by juanje** [comment]
> Hi @juanje! Thank you for working on this. Much appreciated!

Thanks to you, for all the good work you're putting here. I use Toolbox a lot since I installed Silverblue a month ago or so, and I love it :smiling_face_with_three_hearts: 
I'd love to help to make it more robust and useful. Maybe I can't help (yet) much with the Go part, but anything else can help, I will.

> 
> Yes, the `503` errors were my main motivation to make the tests depend on each other because pulling from `registry.fedoraproject.org` is **very** unreliable and causes the CI to flake **a lot**. I think we should contact the registry's maintainers and try to solve this as this is an long-going issue.
> 

I was thinking abut this. I saw it failing now and then with Podman or Buildah, I guess is a known issue, but it must be affecting a lot of people.

> 
> I'm perfectly fine with adding a helper function for pulling the images. Toolbox's image pulling mechanism is simply a wrapper around the `podman pull` command. It is not Toolbox's (or Podman's) fault that `registry.fedoraproject.org` is unreliable.
> 

That's right.
I tried to let **Toolbox** to pull internally the images just at the tests where we really want to test that functionality (e.g., `toolbox create` when there is not toolbox's image downloaded) and call the helper function for the rest.

> 
> I'd love to see more tests but first we need to figure out solutions for some of the problems we're facing now (unreliable registry, #410 - this one might be solved thanks to your update).
> 

I got all the tests working (the registry issue might give some problems, but I think now is more resilient), I'd like to see your thoughts about them (when you have time).
If you like them but have some critics, improvements or anything, just tell me and I'll adjust them.

I add some asserts for the exit status of the commands, to make sure it was the right one. Actually, I found some cases where (I think) **toolbox** should return an exit code `1` (when a command fails), but it returns `0`. I commented this at the tests, so I don't forget but let the tests have similar behavior to the current tests.

For the tests and helper functions I tried to follow this [Shell Style Guide](https://google.github.io/styleguide/shellguide.html) and use `ShellCheck`, so the code is clear, consistent and easy maintain. But if you prefer another style, I can adapt it.

> Also, in the future, we'd like to hook up Toolbox's tests with Podman's. We still have to discuss this matter with Podman developers to get more info about this, though.
> 
> Again, thank you for working on this!

Thanks to you @HarryMichal for take the time to review this and reply :smile:

File: test/system/README.md
Comment (HarryMichal) [inline test/system/README.md:24]: Typo: `btween` -> `between`

File: test/system/rmi.bats
Comment (HarryMichal) [inline test/system/rmi.bats:31]: Either stick to the "Try to" convention or use "Fail" everywhere instead.

File: test/system/run.bats
Comment (HarryMichal) [inline test/system/run.bats:23]: Try to stick to the "Try to" convention or create a different one if you don't like it.

File: test/system/version.bats
Comment (HarryMichal) [inline test/system/version.bats:12]: Should the test case be named as it is when the command, clearly, is not implemented (I don't know if it ever will be)?

File: test/system/run.bats
Comment (juanje) [reply test/system/run.bats:23]: Actually, I tried to stick to the original tests. I only found 3 tests that were following that convention, so I wasn't sure what to do. When I changed a test's title, I tried to stick with the rest of the titles from that file.

Here is the difference between the current tests and my proposal:
```diff
--- 001-version.lst	
+++ version.lst	
 Output version number using full flag
+Output version number using command

--- 002-help.lst	
+++ help.lst	
 Show usage screen when no command is given
+Show usage screen when command help is given
+Show usage screen when flag --help is given

--- 101-list.lst	
+++ list.lst	
-Run list with zero containers and two images
+Run list with zero containers and images
 Run list with zero containers (-c flag)
 Run list with zero images (-i flag)
+Run list with zero toolbox's containers and images, but other image

--- 102-create.lst	
+++ create.lst	
 Create the default container
-Create a container with a valid custom name ('not-running'; name passed as an argument)
-Create a container with a custom image and name ('running';f29; name passed as an option)
+Create a container with a valid custom name ('not-running')
+Create a container with a custom image and name ('running';f29)
 Try to create a container with invalid custom name

--- 103-list.lst	
+++ list.lst	
-Run list with three containers and two images
+Run list with zero containers and images
+Run list with zero containers (-c flag)
+Run list with zero images (-i flag)
+Run list with zero toolbox's containers and images, but other image

--- 201-run.lst	
+++ run.lst	
+Echo 'Hello World' with no containers created (should fail)
 Echo 'Hello World' inside of the default container
 Echo 'Hello World' inside of the 'running' container
-Stop the 'running' container using 'podman stop'
 Echo 'Hello World' again in the 'running' container after being stopped and exit

--- 301-rm.lst	
+++ rm.lst	
 Try to remove a nonexistent container
 Try to remove the running container 'running'
 Remove the not running container 'not-running'
 Force remove the running container 'running'
-Force remove all remaining containers (only 1 should be left)
+Force remove all remaining containers (with 2 containers created and 1 running)

--- 302-rmi.lst	
+++ rmi.lst	
-Remove all images (2 should be present; --force should not be necessary)
+Remove all images with the default image present
+Fail to remove all images with the default image present and running
+Remove all images with the default image present and running (with flag --force)
```

Maybe it's a good time to convert all those tests to the convention or to define a new one.

File: test/system/README.md
Comment (juanje) [reply test/system/README.md:24]: Thanks. I'll fix it.

File: test/system/version.bats
Comment (juanje) [reply test/system/version.bats:12]: I guess this doesn't make much sense. I was thinking about testing when a non-implemented command is passed, which should raise an error. Also, the possibility of that command be implemented.
**Podman** has both, the flag and the command, so I guessed **Toolbox** would implement it in the future.

BTW, the version functionality doesn't appear at the autocomplete nor at the help or man, so I didn't know for sure the expected behavior.

Anyway, I was thinking to convert the test case into a `non-valid-command` case. So, it assures that if a non-valid-command is used, it exits with code `1`, show a clear explanation and how to get the help.
Basically, what this test case does, but as a `non-valid-commnd` test, insted of `command version` test.
Sounds good?

**Comment by juanje** [comment]
Hi @HarryMichal,

I've been working on some of the feedback you gave me. Also, I did some more changes:
* Refactorized the Ansible playbooks as well. The playbooks were identical (almost) and can be unified.
* Added some tasks for mock the toolbox images and fine tune some current tasks.
* Used *Skopeo* to create a local copy of the images, so they only get pulled once. Then (at the tests) used again *skopeo* to restore the images (once deleted), so we don't depend on the reliability of the servers.
* Modified a bit the test's convention and I change all the titles to follow the convention.
* Moved the test for the `version` command to a `non-existent` command.

There are also more minor changes, improvements and a couple the tests more.
For example, a test for the issue #523. I `skip` it because it was failing and I wanted to get the test working and merged before add test for current bugs, but I didn't want to forget about the issue.

Sorry for give you this amount of changes, I know that maybe is too much for review :disappointed: 
Anyway, thanks for your time.

**Review by HarryMichal — CHANGES_REQUESTED** [review]
I perused the list of changes and found a few things that either need to be changed or discussed. Overall, I like where the PR has gotten and I'm confident this could get merged in the future.

Few more notes:

- I'd probably keep the numbering for the test files (e.g. instead of `create.bats` keep it `102-create.bats`). It gives the power to change the order in which the tests are executed while currently they are executed alphabetically.
- I noticed you removed `helpers.bats` because you started relying on `bats-support` + `bats-assert`. That is definitely fine but I think that file will probably have to stay (even though the content will be different). The reason is that currently, the smart use of Skopeo to cache the images is done purely by Zuul but I think the tests will have to use some helper functions at least in that regard because the tests will most likely be run in different environments. I still need to examine this area more, so this is not exactly a very strong point now :).

Thank you for your continuous work on this! I'll try to check this out on my machine and play with it a bit.

File: playbooks/setup-env.yaml
Comment (juanje) [reply playbooks/setup-env.yaml:8]: Yes, of course. It was just temporary while the registry was too flaky. My plan was to change it back before to merge.

File: test/system/README.md
Comment (juanje) [reply test/system/README.md:22]: I agree that isn't necessary, but I had the impression that you wanted that.

In any case, I'd keep the `[command]:` prefix, so the tests results are more clear.

**Comment by juanje** [comment]
Thanks @HarryMichal for taking the time to review it :smile:

> I perused the list of changes and found a few things that either need to be changed or discussed. Overall, I like where the PR has gotten and I'm confident this could get merged in the future.
> 

Great! I know you have been busy stabilizing the go version, so I wasn't expected much attention to it yet. I've been busy myself starting a new job, so I didn't update the PR for long time. I'll try to find time to update it and include your proposals and comments.

> Few more notes:
> 
> * I'd probably keep the numbering for the test files (e.g. instead of `create.bats` keep it `102-create.bats`). It gives the power to change the order in which the tests are executed while currently they are executed alphabetically.
> 

As long as there are no order dependencies, it's fine with me. I think the tests should be independent and isolated, my issue with the numbers was that the could lead to order dependencies, as they had before.

> * I noticed you removed `helpers.bats` because you started relying on `bats-support` + `bats-assert`. That is definitely fine but I think that file will probably have to stay (even though the content will be different). The reason is that currently, the smart use of Skopeo to cache the images is done purely by Zuul but I think the tests will have to use some helper functions at least in that regard because the tests will most likely be run in different environments. I still need to examine this area more, so this is not exactly a very strong point now :).
> 

Well, actually I moved the file to from `test/system/helpers.bash ` to `test/system/libs/helpers.bash`. It has some helpers for the tests.
I could add another function (or improve the current one) to pull the image if it wasn't been copied with `Skopeo` with `Zuul` before. 

> 
> Thank you for your continuous work on this! I'll try to check this out on my machine and play with it a bit.

I don't know if it's working with the current version, but I've tested this branch in Fedora 31, 32 and 33 (from two months ago...) on virtual machines, it should work.
Tell me if you find something broken. Anyway, I'll try to update the PR.


Thanks for your hard work :smile:

**Comment by juanje** [comment]
@HarryMichal I've rebased from master and made some small changes to get the tests working.
I still need to some of the changes you proposed, but I changed the registry for the official one.

There is a (I think) kind of random error at the `run` tests, but just at the Zuul platform. Locally works fine.
I added some debug to check what is going on. The same test, `run: Try to run sudo (with no password) inside of the default container`, sometimes fails at Fedora 32, but others, works fine at that and fails at 31 or 33.

I think this is the same issue that #410. I hope that the debug info give us some light about it.

**Comment by juanje** [comment]
@HarryMichal BTW, I had to rewrite the git history and `push -f` because I had to resolve some conflicts at the rebase.
So, if you want to test it locally, better discard the old branch and pull this version.

**Comment by HarryMichal** [comment]
I took the liberty of rebasing the branch :).

**Comment by hexagonrecursion** [comment]
I want to help move this forward. Here my first patch: https://github.com/juanje/toolbox/pull/1

**Comment by hexagonrecursion** [comment]
Second patch https://github.com/juanje/toolbox/pull/2

**Comment by hexagonrecursion** [comment]
While working on https://github.com/containers/toolbox/issues/630 I added two more tests: https://github.com/hexagonrecursion/toolbox/compare/system-tests-typo...hexagonrecursion:tests-for-issue630?expand=1 I'll open a PR for this branch later.

**Comment by juanje** [comment]
Thanks @hexagonrecursion for working on it. I'll rebase master again to have the pull request updated, I'll add your patches and I'll check that everything work, so we can move forward with the PR.

BTW, about the TODO you put at your first patch, I don't think there is any other way to test the containers. The goal of destroy all the previous containers is to isolate the tests (between them and with external conditions), so they are consistent and we are sure that we are testing the things we think we're testing.

These are system integration tests and I don't think the should be executed on production systems, but on dev/testing (probably virtual) systems, because we need to know that basic functionalities like create and remove containers really works, without rely on other tests, the order of the tests or external conditions.

I personally use Vagrant to spin up different versions of Fedora, run the Ansible playbooks (a slightly modified version of them) to compile Toolbox and run the tests. That way I run the tests on the different versions and in isolation.
Maybe I might update the Vagrant config and put it somewhere so anyone can use it for local testing...

File: test/system/rmi.bats
Comment (juanje) [inline test/system/rmi.bats:18]: It didn't have the assert because the `setup()` function makes sure that the all the images are removed before each test, so this always be 0.
Although, it's probably good idea (for clarity and consistency sake) to add the assert, thanks.

One thing, I do think it's better to use `assert_equal` instead, as @hexagonrecursion suggested. It's a bit clearer and more consistent with the other asserts.
```bash
asser_equal "$num_of_images" 0
```

**Comment by juanje** [comment]
@HarryMichal I rebased the branch with master and pused the changes from @hexagonrecursion , but I can't see the results from Zuul. There is something wrong with the certificate, I don't know if you can see what is failing.

I'll try to update my local environment and test it, to see what is happening and fix it.

BTW, did you manage to get the bats libs's packages into Fedora?

**Comment by juanje** [comment]
Well, I can't see the results yet at Zuul, but I tested locally and fixed the failing tests. Now the branch is in sync with `master`, with the changes proposed by @hexagonrecursion and fix one test that was failing.

**Comment by juanje** [comment]
@HarryMichal I just saw this error, after rebase master:

```
regexp : ^toolbox version [0-9]+\.[0-9]+\.[0-9]+$
output : toolbox version 0.0.98.1
```
Did you change the version format (four numbers instead of tree)? Is it a new format or just a one time thing?
If this is a new format I'll change the test.

Thanks.

**Comment by juanje** [comment]
@HarryMichal I added a test for the new `--distro` option.

Also, checking the error at the [last Zuul's check](https://softwarefactory-project.io/zuul/t/local/build/a78f4e640fdc46aebcf7635c3cfa3546/console) (which is the random error that sometimes pop up) the output show this:
```
Error: container create failed (no logs from conmon): EOF
```
It's weird, because the container seems to exist, but that message says otherwise...

I'm not able to reproduce this error locally, with VM or at Cloud instance and it isn't happening always at Zuul, just once in a while. It seems like Zuul's environment issue, but it's hard to debug, even with the extra debug I put there...

Anyway, maybe you find something else from the logs.

File: playbooks/fedora-31/setup-env.yaml
Comment (debarshiray) [inline playbooks/fedora-31/setup-env.yaml]: Nitpick: shouldn't the `repo`, `depth` and `dest` lines be indented? Because they are arguments for `git`.

Comment (HarryMichal) [reply playbooks/fedora-31/setup-env.yaml]: Yes, this is a mistake I made when squashing commits. Fixed.

**Comment by HarryMichal** [comment]
> @HarryMichal I added a test for the new `--distro` option.

Awesome!

> Also, checking the error at the [last Zuul's check](https://softwarefactory-project.io/zuul/t/local/build/a78f4e640fdc46aebcf7635c3cfa3546/console) (which is the random error that sometimes pop up) the output show this:
> 
> ```
> Error: container create failed (no logs from conmon): EOF
> ```
> 
> It's weird, because the container seems to exist, but that message says otherwise...
> 
> I'm not able to reproduce this error locally, with VM or at Cloud instance and it isn't happening always at Zuul, just once in a while. It seems like Zuul's environment issue, but it's hard to debug, even with the extra debug I put there...
> 
> Anyway, maybe you find something else from the logs.

The error is certainly weird... It seems there have been some reports with a similar error message. I went through some but none seem to be related much. Weird thing is that this showed up now and not in the past, so we kinda can say, that these changes caused it. But is the true core of the problem... that I don't know.

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
recheck

**Comment by juanje** [comment]
> > @HarryMichal I added a test for the new `--distro` option.
> 
> Awesome!
> 

I think that the new test for `--distro` got lost at the squashing. I don't see it at the `101-create.bats` :thinking: 

In case you lost it and you want to put it back, here is the code:

```
@test "create: Create a container with a distro and release options ('fedora'; f29)" {
  pull_image 29

  run toolbox -y create -d "fedora" -r f29

  assert_success
  assert_output --partial "Created container: fedora-toolbox-29"
  assert_output --partial "Enter with: toolbox enter --release 29"

  # Make sure the container has actually been created
  run podman ps -a

  assert_output --regexp "Created[[:blank:]]+fedora-toolbox-29"
}

```

**Comment by HarryMichal** [comment]
> > > @HarryMichal I added a test for the new `--distro` option.
> > 
> > 
> > Awesome!
> 
> I think that the new test for `--distro` got lost at the squashing. I don't see it at the `101-create.bats` thinking
> 
> In case you lost it and you want to put it back, here is the code:
> 
> ```
> @test "create: Create a container with a distro and release options ('fedora'; f29)" {
>   pull_image 29
> 
>   run toolbox -y create -d "fedora" -r f29
> 
>   assert_success
>   assert_output --partial "Created container: fedora-toolbox-29"
>   assert_output --partial "Enter with: toolbox enter --release 29"
> 
>   # Make sure the container has actually been created
>   run podman ps -a
> 
>   assert_output --regexp "Created[[:blank:]]+fedora-toolbox-29"
> }
> ```

@juanje Aaah!! I messed up! I forgot to cherry-pick the commit into my local copy where I had the branch squashed already from December. Thank you for posting the code here. Apologies!

**Comment by HarryMichal** [comment]
I took the liberty of slightly adjusting the test case. Instead of relying on `podman ps` to get a list of containers, use `toolbox list -c`.

**Comment by juanje** [comment]
> I took the liberty of slightly adjusting the test case. Instead of relying on `podman ps` to get a list of containers, use `toolbox list -c`.

At first I thought to do the same, but later I thought that testing a `toolbox`'s command with another `toolbox`'s command was such a good idea. Then you depend on other `toolbox list -c` to be working fine, which is kind of inter-dependency between tests. If (for whatever reason) the command `list` changes or breaks, this test breaks.

It's easier/cleaner to test against the `list` command output, but I think it's better to avoid this kind of inter-dependencies.
Just a thought.

**Comment by HarryMichal** [comment]
> > I took the liberty of slightly adjusting the test case. Instead of relying on `podman ps` to get a list of containers, use `toolbox list -c`.
> 
> At first I thought to do the same, but later I thought that testing a `toolbox`'s command with another `toolbox`'s command was such a good idea. Then you depend on other `toolbox list -c` to be working fine, which is kind of inter-dependency between tests. If (for whatever reason) the command `list` changes or breaks, this test breaks.
> 
> It's easier/cleaner to test against the `list` command output, but I think it's better to avoid this kind of inter-dependencies.
> Just a thought.

You've got a point here and I'm sorry for doubting your reasoning behind this.

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
recheck

**Comment by HarryMichal** [comment]
Seems like the flakiness went away. I'm still not sure what it was caused, though. So, I say let's merge this!

What do you think, @debarshiray?

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for rebasing everything!

File: test/system/README.md
Comment (debarshiray) [inline test/system/README.md:25]: I merged this part via https://github.com/containers/toolbox/pull/700

File: playbooks/toolbox-mock-images.yaml
Comment (juanje) [reply playbooks/toolbox-mock-images.yaml:23]: `list` is a filter:
https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#combining-items-from-multiple-lists-zip-and-zip-longest

File: .zuul.yaml
Comment (juanje) [reply .zuul.yaml:21]: Right now there are no package conflicts and the `setup-test-{fedora-version}.yml` are identical. I think that it can be a source of future issues (not updating all the files), also unnecessary duplication.
If some RPM packages change in the future, they can be manage easily using variables inside the playbook, as there will be just a few, but the rest of the playbook will be identical.

File: playbooks/pre-common.yaml
Comment (juanje) [reply playbooks/pre-common.yaml:3]: Actually, we don't. I added this before to unify the `setup-env.yaml`, to avoid the duplication of this. But if we unify the `setup-env.yml` we could move this code into that file.

File: .zuul.yaml
Comment (debarshiray) [reply .zuul.yaml:21]: Ok, then. :)

Just note that I know very little about Ansible, and even less about using variables, etc.. So, I might end up having to ask for your help.

File: playbooks/pre-common.yaml
Comment (debarshiray) [reply playbooks/pre-common.yaml:3]: Ok, understood.

File: .zuul.yaml
Comment (debarshiray) [reply .zuul.yaml:21]: Although, we should probably still retain the `fedora` sub-directory, because we might be adding support for other distributions like Ubuntu.

File: playbooks/toolbox-mock-images.yaml
Comment (HarryMichal) [reply playbooks/toolbox-mock-images.yaml:7]: Good catch. Forgot to update this.

Comment (debarshiray) [reply playbooks/toolbox-mock-images.yaml:7]: It's actually a bit tricky to use the new locations of the images because the Fedora 29 image is still at the old location. I think I managed to get it to work by open-coding the loop and the variables. See https://github.com/containers/toolbox/pull/701

**Comment by debarshiray** [comment]
The rebase against current `master` was a bit more involved than I expected. Given my general lack of Ansible skills, I split things up a bit into smaller PRs to be sure that I didn't mess something up:
* https://github.com/containers/toolbox/pull/702
* https://github.com/containers/toolbox/pull/703

File: playbooks/toolbox-mock-images.yaml
Comment (debarshiray) [reply playbooks/toolbox-mock-images.yaml:23]: I see. Thanks for the link, @juanje !

In the end, I had to [open-code the loops](https://github.com/containers/toolbox/pull/704) because all the other images, except the one for Fedora 29, have moved to a newer path on `registry.fedoraproject.org`, and I couldn't trust my Ansible skills to properly parametrize that.

We could either wait for a few more months when Fedora 32 gets EOLed, and we can use the F32 image instead of the F29 one and all the URLs will be in the same pattern again. Or we could use your help to improve things now in a separate PR. :)

**Review by debarshiray — APPROVED** [review]

**Comment by debarshiray** [comment]
recheck

**Comment by debarshiray** [comment]
recheck

**Comment by debarshiray** [comment]
Thanks for all the work on this, @juanje & @HarryMichal ! Much appreciated.

**Comment by HarryMichal** [comment]
All the glory goes to @juanje. He did an amazing job! :tada:

**Comment by juanje** [comment]
Thanks to both @debarshiray and @HarryMichal for letting me contribute my grain of sand to this project. You're doing a great job!
I'll try to help any way I can to the project. I'll be happy to add more test to improve the coverage or keep the current ones updated now that this is been merged :smile:

I think this project is great for developers and for normal users that like the terminal.

**Pull request merged**
